src/lib.rs

```diff
@@ -52,18 +52,70 @@ pub fn multi_index_map(input: proc_macro::TokenStream) -> proc_macro::TokenStrea
             }
             Uniqueness::NonUnique => match ordering {
                 Ordering::Hashed => quote! {
-                    #index_name: rustc_hash::FxHashMap<#ty, Vec<usize>>,
+                    #index_name: rustc_hash::FxHashMap<#ty, std::collections::BTreeSet<usize>>,
                 },
                 Ordering::Ordered => quote! {
-                    #index_name: std::collections::BTreeMap<#ty, Vec<usize>>,
+                    #index_name: std::collections::BTreeMap<#ty, std::collections::BTreeSet<usize>>,
                 }
             }
         }
     });
 
+    // For each indexed field generate a TokenStream representing initializing the lookup table. Used in `with_capacity` initialization
+    // If lookup table data structures support `with_capacity`, change `default()` and `new()` calls to `with_capacity(n)`
+    let lookup_table_fields_init: Vec<proc_macro2::TokenStream> = fields_to_index().map(|f|{
+        let index_name = format_ident!("_{}_index", f.ident.as_ref().unwrap());
+        let (ordering, _uniqueness) = get_index_kind(f).unwrap_or_else(|| {
+            abort_call_site!("Attributes must be in the style #[multi_index(hashed_unique)]")
+        });
+        match ordering {
+            Ordering::Hashed => quote! {
+                #index_name: rustc_hash::FxHashMap::default(),
+            },
+            Ordering::Ordered => quote! {
+                #index_name: std::collections::BTreeMap::new(),
+            }
+        }
+    }).collect();
+
+    // For each indexed field generate a TokenStream representing reserving capacity in the lookup table. Used in `reserve` 
+    // If the Ordered lookup table data structure support `reserve`, it should also call `reserve(additional)`
```

Comment:
Given how btree works, at this moment I don't see a way to extend its capacity that does not involve adding dummy nodes (which may save time from memory allocation but makes the insert and delete time complexity worse). It will be interesting to see how it is done (currently, BtreeMap does not implement extend_reserve(), and the trait default impl does nothing).

Reply:
[2bff67e](https://github.com/lun3x/multi_index_map/pull/23/commits/2bff67eec2587697fc6088cd8e6a6bd722e992e2)